### PR TITLE
Method Voronoi

### LINKED ISSUE #160: Method D (Voronoi) | Calculation of Voronoi velocity
**JuPedSim - JPSreport**

- Version        : 0.8.4
- Commit hash    : 74db190

**Problem description**
When calculating the Voronoi density and Voronoi speed with the current version I can't get the results of [J. Zhang 2012, p. 66, figure 4.6](http://hdl.handle.net/2128/4898) for *UO-080-300-300* (run of experiment of two-dimensional undirectional movement with open boundary conditions (*UO*), see [Pedestrian Dynamics Data Archive](http://ped.fz-juelich.de/da/2009unidirOpen)).

The comparison of the time series shows that the density differs only minimally:

Method_D_density

When comparing the time series for speed, differences are clearly visible:

Method_D_velocity

I had expected that the time series for the speed should be similar.

**Expected behavior**
I had expected that the time series for velocity and density would be similar to the Method C (Classical):

Method_C_density

Method_C_velocity

The differences could be numerical.

The two methods are currently implemented as follows:

**Method C (Classical)** (`GetClassicVelocity` in [Method_C.cpp, line 115 et seqq.](https://github.com/JuPedSim/jpsreport/blob/81a4d73800e4fc705e018668e9d09467afac5a7c/methods/Method_C.cpp#L115))

For the [classical method](http://www.jupedsim.org/jpsreport/2016-11-06-method-c.html), the spatial mean velocity is the average of the instantaneous velocities for all pedestrians in the measurement area at time:

Clasic_speed

This is implemented in the program code as follows

```cpp
double Method_C::GetClassicalVelocity(const vector<double>& xs, const vector<double>& ys, const vector<double>& VInFrame, int pednum) const
{
     int pedsinMeasureArea=0;
     double velocity = 0;
     for(int i=0; i<pednum; i++)
     {
          if(within(make<point_2d>(xs[i], ys[i]), _areaForMethod_C->_poly))
          {
               velocity+=VInFrame[i];
               pedsinMeasureArea++;
          }
     }
     if(pedsinMeasureArea!=0)
     {
          velocity /= (1.0*pedsinMeasureArea);
     } else {
          velocity = 0;
     }
     return velocity;

}
```

This function was not edited in the last release.

**Method D (Voronoi)** (`GetVoronoiDensityVelocity` in [Method_D.cpp, line 326 et seqq.](https://github.com/JuPedSim/jpsreport/blob/81a4d73800e4fc705e018668e9d09467afac5a7c/methods/Method_D.cpp#L326))

For the [Voronoi method](http://www.jupedsim.org/jpsreport/2016-11-07-method-d.html), the speed (*Voronoi velocity*) for the measurement area is defined as:

Voronoi_speed

```cpp
std::tuple<double,double> Method_D::GetVoronoiDensityVelocity(const vector<polygon_2d>& polygon, const vector<double>& Velocity, const polygon_2d & measureArea)
{
     double meanV=0;
     double density=0;
     int temp=0;
     for (auto && polygon_iterator:polygon)
     {
          polygon_list v;
          intersection(measureArea, polygon_iterator, v);
          if(!v.empty())
          {
               meanV+=Velocity[temp]*area(v[0]);
               density+=area(v[0])/area(polygon_iterator);
               if((area(v[0]) - area(polygon_iterator))>J_EPS)
               {
                    std::cout<<"----------------------Now calculating density-velocity!!!-----------------\n ";
                    std::cout<<"measure area: \t"<<std::setprecision(16)<<dsv(measureArea)<<"\n";
                    std::cout<<"Original polygon:\t"<<std::setprecision(16)<<dsv(polygon_iterator)<<"\n";
                    std::cout<<"intersected polygon: \t"<<std::setprecision(16)<<dsv(v[0])<<"\n";
                    std::cout<<"this is a wrong result in density calculation\t "<<area(v[0])<<'\t'<<area(polygon_iterator)<<  "  (diff=" << (area(v[0]) - area(polygon_iterator)) << ")" << "\n";
               }
          }
          temp++;
     }
     meanV = meanV/area(measureArea);
     density = density/(area(measureArea)*CMtoM*CMtoM);
     return std::make_tuple(density, meanV);
}
```

This function was edited in the last release (commits 694e497743ae5ae0f3656100d763e684858f7ae6 and ec35c62d3ccc2687b5f8cabe4c992f956065bd14).

**Conclusion**

We had agreed in issue #127 ([gitlab issue 84](https://gitlab.version.fz-juelich.de/jupedsim/jpsreport/issues/84)) that 

> - Method D implements the abovementioned formula correctly. So the code is not buggy.
> - When to use method D and when not is an important question. I hope it can be discussed in a publication. However, a short notice to warn users can be written in the documentation of method D.
> - Other enhancements or changes of these formulae can be implemented. I would suggest a different method for that.

I just can't explain why I can't get the results from Zhang anymore.

@chraibi Were we wrong?

### PR DESCRIPTION
**Pull request by gjaeger**
This proposal resolves #160.  The problem of velocity calculation in method D is solved by a combination of classical velocity calculation and Voronoi density calculation. This was implemented in the new method _Voronoi_.

The differences between the methods _C_, _D_ and _Voronoi_ are as follows (Utest/Methods/ini_Methods.xml):

v-t

The difference between methods _C_ and _Voronoi_ (V) is explained by the consideration of pedestrians who have an intersection with the measurement area in method _Voronoi_ when calculating the mean velocity.

### CODE REVIEW

File: Analysis.cpp [hunk 1]
```diff
@@ -77,6 +78,7 @@ Analysis::Analysis()
      _DoesUseMethodC = false;                                   // Method C //calculate and save results of classic in separate file
      _DoesUseMethodD = false;                                   // Method D--Voronoi method
      _DoesUseMethodI = false;
+     _DoesUseMethod_Voronoi = false;
```
Comment (chraibi) [inline Analysis.cpp]: why this variable has a different form than previous ones?

File: Utest/Methods/ini_Methods.xml [hunk 2]
```diff
@@ -71,22 +69,28 @@
...
     <profiles enabled="false" grid_size_x="0.20" grid_size_y="0.20"/>
-	</method_D>
+  </method_D>
 
   <method_I enabled="true">
     <measurement_area id="1" start_frame="None" stop_frame="None" get_individual_FD="true"/>
     <cut_by_circle enabled="true" radius="1.0" edges="10"/>
 	</method_I>
 
+  <method_Voronoi enabled="true">
```
Comment (chraibi) [inline Utest/Methods/ini_Methods.xml]: is `method_voronoi` the same as `method_V`?
if yes, why this naming is different?
if no, why what's the difference between both methods?

File: methods/Method_Voronoi.cpp [hunk 3]
```diff
@@ -0,0 +1,720 @@
+/**
+ * \file        Method_Voronoi.cpp
+ * \date        Oct 10, 2014
```
Comment (chraibi) [inline methods/Method_Voronoi.cpp]: change date and version

File: methods/Method_Voronoi.cpp [hunk 4]
```diff
@@ -0,0 +1,720 @@
...
+               }
+          }
+          i++;
+     }
+
+     if(pedsinMeasureArea!=0)
+      {
+        velocity /= (1.0*pedsinMeasureArea);
+      } else {
+        velocity = 0;
```
Comment (chraibi) [inline methods/Method_Voronoi.cpp]: no need for this `else`, since `velocity` is initialised to 0.

Comment (gjaeger) [reply methods/Method_Voronoi.cpp]: solved with 2da4fd5.

File: methods/Method_Voronoi.cpp (hunk 3 above)
Comment (gjaeger) [reply methods/Method_Voronoi.cpp]: solved with 2da4fd5.

File: methods/Method_Voronoi.h [hunk 5]
```diff
@@ -0,0 +1,112 @@
+/**
+ * \file        Method_Voronoi.h
+ * \date        Oct 10, 2014
```
[the remark this answers is not in the document]
Comment (gjaeger) [reply methods/Method_Voronoi.h]: solved with 2da4fd5.

File: Analysis.cpp (hunk 1 above)
Comment (gjaeger) [reply Analysis.cpp]: solved with 72dba92.

File: Utest/Methods/ini_Methods.xml (hunk 2 above)
Comment (gjaeger) [reply Utest/Methods/ini_Methods.xml]: solved with 72dba92.

**Pull request merged**